examples/ex42.cpp

```diff
@@ -1101,31 +1217,64 @@ static inline void AssembleGradientDiagonal3D(
    auto Ke_diag_m =
       Reshape(Ke_diag_memory.ReadWrite(), d1d, d1d, d1d, dim, ne, dim);
 
-   MFEM_FORALL(e, ne,
+   MFEM_FORALL_3D(e, ne, q1d, q1d, q1d,
    {
-      tensor<double, d1d, d1d, d1d, dim, dim> Ke_diag{};
-
+      // shared mememory 1D Basis functions, filled from b and g inputs
+      MFEM_SHARED tensor<double, q1d, d1d> B, G;
+      // shared memory placeholders for temporary contraction results
+      MFEM_SHARED tensor<double, 2, 3, q1d, q1d, q1d> smem;
       // du/dxi
-      tensor<double, q1d, q1d, q1d, dim, dim> dudxi{};
+      MFEM_SHARED tensor<double, q1d, q1d, q1d, dim, dim> dudxi;
+      MFEM_SHARED tensor<double, d1d, d1d, d1d, dim, dim> Ke_diag;
+
+      // B, G, dudxi, Ke_diag initialization
+      MFEM_FOREACH_THREAD(qx, x, q1d)
+      {
+         MFEM_FOREACH_THREAD(qy, y, q1d)
+         {
+            if (MFEM_THREAD_ID(z) == 0 && qy < d1d)
+            {
+               const int dy = qy;
+               B(qx, dy) = b(qx, dy);
+               G(qx, dy) = g(qx, dy);
+            }
+            MFEM_FOREACH_THREAD(qz, z, q1d)
+            {
+               const bool inside_d1d_cube = qx < d1d && qy < d1d && qz < d1d;
+               for (int l = 0; l < dim; l++)
+               {
+                  for (int m = 0; m < dim; m++)
+                  {
+                     dudxi(qx, qy, qz, l, m) = 0.0;
+                     if (inside_d1d_cube) { Ke_diag(qx, qy, qz, l,m) = 0.0; }
+                  }
+               }
+            }
```

Comment:
same comment as https://github.com/mfem/mfem/pull/2874/files#r822201282

is there an implicit assumption being made that q1d will never be less than d1d? I understand you don't attain the maximum-rank element matrices by such integration rules, but it still is pretty common.

Reply:
edit: @camierjs showed me that there is a check to ensure this assumption holds:

```
   MFEM_VERIFY(d1d <= q1d, "d1d <= q1d");
```

but it's in a different part of the code, so I missed it!